src/darwin/Framework/CHIP/MTREndpointInfo.mm

```diff
@@ -0,0 +1,210 @@
+    // PartsList we can assign it as a child of the endpoint we're processing, and we can be sure
+    // that this is the closest parent, not some higher ancestor using the Full-Family Pattern.
+    NSMutableArray<MTREndpointInfo *> * children = [[NSMutableArray alloc] init];
+    while (!deque.empty()) {
+        EndpointId endpointID = deque.front();
+        MTREndpointInfo * endpoint = endpoints[@(endpointID)];
+        deque.pop_front();
+
+        if (endpoint->_mark == EndpointMark::ParentAssigned) {
+            continue; // This endpoint is part of a cycle, don't populate its children.
+        }
+
+        [children removeAllObjects];
+        for (NSNumber * partNumber in endpoint->_partsList) {
+            MTREndpointInfo * descendant = endpoints[partNumber];
+            if (descendant != nil && descendant->_mark != EndpointMark::ParentAssigned) {
+                descendant->_mark = EndpointMark::ParentAssigned;
+                [children addObject:descendant];
+            }
+        }
+        endpoint->_children = [children copy];
+    }
+    root->_mark = EndpointMark::ParentAssigned;
+    return valid;
+}
+
++ (NSDictionary<NSNumber *, MTREndpointInfo *> *)endpointsFromAttributeCache:(const ClusterStateCache *)cache
+{
+    VerifyOrReturnValue(cache != nullptr, nil);
+    using namespace Descriptor::Attributes;
+
+    NSMutableDictionary<NSNumber *, MTREndpointInfo *> * endpoints = [[NSMutableDictionary alloc] init];
+    cache->ForEachAttribute(Descriptor::Id, [&](const ConcreteAttributePath & path) -> CHIP_ERROR {
+        VerifyOrReturnError(path.mAttributeId == DeviceTypeList::Id, CHIP_NO_ERROR);
+
+        CHIP_ERROR err = CHIP_NO_ERROR;
+        NSArray<MTRDescriptorClusterDeviceTypeStruct *> * deviceTypeList = MTRDecodeAttributeValue(path, *cache, &err);
+        if (!deviceTypeList) {
+            MTR_LOG_ERROR("Ignoring endpoint %u, failed to parse DeviceTypeList: %" CHIP_ERROR_FORMAT, path.mEndpointId, err.Format());
+            return CHIP_NO_ERROR;
+        }
+
+        NSMutableArray * deviceTypes = [[NSMutableArray alloc] initWithCapacity:deviceTypeList.count];
+        for (MTRDescriptorClusterDeviceTypeStruct * deviceTypeStruct in deviceTypeList) {
+            MTRDeviceTypeRevision * type = [[MTRDeviceTypeRevision alloc] initWithDeviceTypeStruct:deviceTypeStruct];
+            if (!type) {
+                MTR_LOG_ERROR("Ignoring invalid device type 0x%x rev %u for endpoint %u",
+                    deviceTypeStruct.deviceType.unsignedIntValue, deviceTypeStruct.revision.unsignedShortValue,
+                    path.mEndpointId);
+                continue;
+            }
+            [deviceTypes addObject:type];
+        }
+        if (!deviceTypes) {
+            MTR_LOG_ERROR("Ignoring endpoint %u, no device types", path.mEndpointId);
+            return CHIP_NO_ERROR;
+        }
+
+        ConcreteAttributePath partsListPath(path.mEndpointId, path.mClusterId, PartsList::Id);
```

Comment:
So if the device is buggy and has no DeviceTypeList attribute, we won't populate the PartsList bits too.  Might be ok, but worth at least documenting clearly.

Reply:
I've changed the code to just carry on in this case, we should probably just carry on in a best-effort fashion.